plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/AsyncUploadUtils.java

```diff
@@ -0,0 +1,366 @@
+            () -> s3AsyncClient.uploadPart(
+                uploadPartRequest,
+                AsyncRequestBody.fromInputStream(stream.getInputStream(), stream.getContentLength(), streamReadExecutor)
+            )
+        );
+
+        CompletableFuture<CompletedPart> convertFuture = uploadPartResponseFuture.thenApply(
+            uploadPartResponse -> convertUploadPartResponse(completedParts, uploadPartResponse, partNumber)
+        );
+        futures.add(convertFuture);
+
+        CompletableFutureUtils.forwardExceptionTo(convertFuture, uploadPartResponseFuture);
+    }
+
+    private CompletedPart convertUploadPartResponse(
+        AtomicReferenceArray<CompletedPart> completedParts,
+        UploadPartResponse partResponse,
+        int partNumber
+    ) {
+        CompletedPart completedPart = CompletedPart.builder().eTag(partResponse.eTag()).partNumber(partNumber).build();
+        completedParts.set(partNumber - 1, completedPart);
+        return completedPart;
+    }
+
+    /**
+     * Calculates the optimal part size of each part request if the upload operation is carried out as multipart upload.
+     */
+    public long calculateOptimalPartSize(long contentLengthOfSource) {
+        if (contentLengthOfSource < ByteSizeUnit.MB.toBytes(5)) {
+            return contentLengthOfSource;
+        }
+        double optimalPartSize = contentLengthOfSource / (double) MAX_UPLOAD_PARTS;
+        optimalPartSize = Math.ceil(optimalPartSize);
+        return (long) Math.max(optimalPartSize, minimumPartSize);
+    }
+
+    private void uploadInOneChunk(
+        S3AsyncClient s3AsyncClient,
+        UploadRequest uploadRequest,
+        Stream stream,
+        CompletableFuture<UploadResponse> returnFuture
+    ) {
+        PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+            .bucket(uploadRequest.getBucket())
+            .key(uploadRequest.getKey())
+            .contentLength(uploadRequest.getContentLength())
+            .build();
+        ExecutorService streamReadExecutor = uploadRequest.getWritePriority() == WritePriority.HIGH
+            ? priorityExecutorService
+            : executorService;
+        CompletableFuture<UploadResponse> putObjectFuture = SocketAccess.doPrivileged(
+            () -> s3AsyncClient.putObject(
+                putObjectRequest,
+                AsyncRequestBody.fromInputStream(stream.getInputStream(), stream.getContentLength(), streamReadExecutor)
+            ).handle((resp, throwable) -> {
+                if (throwable != null) {
+                    returnFuture.completeExceptionally(throwable);
+                } else {
+                    uploadRequest.getUploadFinalizer().accept(true);
+                    returnFuture.complete(new UploadResponse(true));
+                }
+
+                return null;
+            }).handle((resp, throwable) -> {
+                if (throwable != null) {
+                    deleteUploadedObject(s3AsyncClient, uploadRequest);
```

Comment:
This deletion is in cases where `UploadFinalizer#finalizeUpload` fails. Currently we are only doing local data integrity checks in this method. If local data integrity checks fail, we need to delete the uploaded object in case of single chunk upload and abort upload in case of multipart uploads. That is what this handling is achieving.